samples: shields: lmp90100_evb: convert to using DEVICE_DT_GET_ONE()

henrikbrixandersen opened the pull request:
Convert from using device_get_binding() to DEVICE_DT_GET_ONE().

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

avisconti reviewed (approved):
LGTM

gmarull commented on samples/shields/lmp90100_evb/rtd/src/main.c:
```diff
@@ -60,7 +60,7 @@ static double rtd_temperature(int nom, double resistance)
 void main(void)
 {
-	const struct device *lmp90100;
+	const struct device *lmp90100 = DEVICE_DT_GET(DT_INST(0, ti_lmp90100));
```
`DEVICE_DT_GET_ANY(ti_lmp90100)` is a better choice for samples

henrikbrixandersen replied:
`DEVICE_DT_GET(DT_INST(0, ti_lmp90100))` results in a compilation error if no instance is enabled, `DEVICE_DT_GET_ANY(ti_lmp90100)` results in a runtime error.

mbolivar-nordic replied:
DEVICE_DT_GET_ANY is in fact a better choice for samples. We do not want to use DT_INST in samples. https://github.com/zephyrproject-rtos/zephyr/issues/32139

henrikbrixandersen replied:
Why? Dropping it will make the sample compile even if misconfigured and only fail at runtime, thus not catching the error in CI (unless ran on a supporting configuration, which I doubt we have in any board cluster).

What is the reason behind your request?

mbolivar-nordic replied:
It is explained in some detail in the issue I linked.

henrikbrixandersen replied:
I'm sorry, but I cannot find it in the issue, you linked. I see the argument that some find it confusing, but also ghat you would bring it up on the API meeting before converting samples. 

Has it been brought up yet? I don't agree with designing for runtime errors instead of compile time errors, especially not for samples, which we compile in CI for a reason.

mbolivar-nordic replied:
> I'm sorry, but I cannot find it in the issue, you linked.

The rationale for removing DT_INST from samples is right there in the issue description.

>  I see the argument that some find it confusing, but also that you would bring it up on the API meeting before converting samples. Has it been brought up yet?

Where did I say I would bring it up in the API meeting? I don't see that mentioned anywhere.

The issue was brought up at dev-review, however, and the consensus was to move forward with the removal of DT_INST from samples as documented in that issue. Here are the minutes:

https://docs.google.com/document/d/15KUINNj7Re0GQiWYpqKvO5jgDw8oiQv3SGJfNHGaPOU/edit#heading=h.1ylyjgx28i5t

(Scroll to 11 February 2021 minutes if that heading anchor doesn't work.)

> I don't agree with designing for runtime errors instead of compile time errors, especially not for samples, which we compile in CI for a reason.

OK, I am sympathetic to that, but that is not an argument for continuing to use DT_INST in a sample.

I guess we could add something like this instead:

```c
#define DEVICE_DT_GET_ANY_OR_COMPILE_ERROR(compat)			    \
	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(compat),			    \
		    (DEVICE_DT_GET(DT_COMPAT_GET_ANY_STATUS_OKAY(compat))), \
		    (((const struct device*)NULL) +                         \
                     ZERO_OR_COMPILE_ERROR(0)))
```

mbolivar-nordic reviewed (changes requested):
Please drop the DT_INST

henrikbrixandersen committed:
device: add DEVICE_DT_GET_ONE macro

Add DEVICE_DT_GET_ONE() macro. This macro is similar to
DEVICE_DT_GET_ANY() but instead of returning NULL if no enabled,
compatible device instance is found, it throws a compilation error.

This is useful for compile-time checks in samples.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
samples: shields: lmp90100_evb: convert to using DEVICE_DT_GET_ONE()

Convert from using device_get_binding() to DEVICE_DT_GET_ONE().

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

mbolivar-nordic reviewed (approved):
Looks good. If you have time, please fix the commit message to say "[...] convert to using DEVICE_DT_GET_ONE"

henrikbrixandersen commented:
> Looks good. If you have time, please fix the commit message to say "[...] convert to using DEVICE_DT_GET_ONE"

@mbolivar-nordic  Thanks. I have updated the commit message.

gmarull commented on samples/shields/lmp90100_evb/rtd/src/main.c:
```diff
@@ -60,7 +60,7 @@ static double rtd_temperature(int nom, double resistance)
 void main(void)
 {
-	const struct device *lmp90100;
+	const struct device *lmp90100 = DEVICE_DT_GET_ONE(ti_lmp90100);
```
Just a quick question: what should samples use now, `DEVICE_DT_GET_ONE` or `DEVICE_DT_GET_ANY`? It may be confusing for new users to have both.

mbolivar-nordic replied:
Sigh. I was hoping nobody would ask that.

I don't know if there is a best answer. DEVICE_DT_GET_ANY gives better error messages, but they're at runtime. DEVICE_DT_GET_ONE fails fast, but the error message when there's no such device is very confusing to beginners.
